File: Example/ActionsListDemoObjectiveC/ViewController.m
//
//  ActionsList
//  Copyright © 2018 LowKostKustomz. All rights reserved.
//

#import "ViewController.h"
@import ActionsList;

@interface ViewController ()

@end

@implementation ViewController {
    ActionsListModel* list;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.navigationItem setTitle:NULL];
    [self createBackground];
    [self createNavigationItems];
    
    UIButton* button = [UIButton buttonWithType:UIButtonTypeSystem];
    [button setImage:[[UIImage imageNamed:@"Clock app"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
            forState:UIControlStateNormal];
    [button addTarget:self action:@selector(clockButtonAction:) forControlEvents:UIControlEventTouchUpInside];
    [button setAdjustsImageWhenHighlighted:NO];
    
    [self.view addSubview:button];
    [button setTranslatesAutoresizingMaskIntoConstraints:NO];
    [[button.heightAnchor constraintEqualToConstant:60] setActive:YES];
    [[button.widthAnchor constraintEqualToAnchor:button.heightAnchor] setActive:YES];
    
    [[button.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor constant:-64] setActive:YES];
    [[button.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor constant:-100] setActive:YES];
    [button setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisHorizontal];
}

- (void)createBackground {
    UIImageView* imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"Wallpaper"]];
    [imageView setContentMode:UIViewContentModeScaleAspectFill];
    
    [self.view addSubview:imageView];
    [imageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [[imageView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[imageView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    [[imageView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor] setActive:YES];
    [[imageView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor] setActive:YES];
}

- (void)createNavigationItems {
    UIButton* rightButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [rightButton setTitleColor:UIColor.blackColor forState:UIControlStateNormal];
    [rightButton setTitle:@"Button" forState:UIControlStateNormal];
    [rightButton sizeToFit];
    [rightButton addTarget:self action:@selector(presentFromNavigation:) forControlEvents:UIControlEventTouchUpInside];
    
    UIButton* rightButton1 = [UIButton buttonWithType:UIButtonTypeCustom];
    [rightButton1 setImage:[UIImage imageNamed:@"List icon"] forState:UIControlStateNormal];
    [rightButton1 setAdjustsImageWhenHighlighted:NO];
    [rightButton1 setTintAdjustmentMode:UIViewTintAdjustmentModeNormal];
    [rightButton1 setTintColor:UIColor.blackColor];
    [rightButton1 sizeToFit];
    [rightButton1 addTarget:self action:@selector(presentFromNavigation:) forControlEvents:UIControlEventTouchUpInside];
    
    [self.navigationItem setRightBarButtonItems:@[[[UIBarButtonItem alloc] initWithCustomView:rightButton1],
                                                  [[UIBarButtonItem alloc] initWithCustomView:rightButton]]];
    
    [self.navigationItem setLeftBarButtonItems:@[[[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"List icon"]
                                                                                  style:UIBarButtonItemStylePlain
                                                                                 target:self
                                                                                 action:@selector(presentFromNavigationBar:)],
                                                 [[UIBarButtonItem alloc] initWithTitle:@"BarButton"
                                                                                  style:UIBarButtonItemStylePlain
                                                                                 target:self
                                                                                 action:@selector(presentFromNavigationBar:)]]];
}

- (void)clockButtonAction:(UIButton*)sender {
    ActionsListModel* model = [sender createActionsListWithDelegate:NULL];
    [model setupDismissAccessibilityHint:@"Dismiss action label here"];
    ActionsListDefaultButtonModel* createAlarmAction = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Create Alarm"
                                                                                                               image:[UIImage imageNamed:@"Alarm clock"]
                                                                                                              action:^(ActionsListDefaultButtonModel* model) {
                                                                                                                  [model.list dismissWithCompletion:NULL];
                                                                                                              }
                                                                                                           isEnabled:YES];
    [createAlarmAction setupAccessibilityWithAccessibilityLabel:@"Create Alarm"];
    [model addAction:createAlarmAction];
    ActionsListDefaultButtonModel* startStopwatchAction = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Start Stopwatch"
                                                                                                                  image:[UIImage imageNamed:@"Stopwatch"]
                                                                                                                 action:^(ActionsListDefaultButtonModel* model) {
                                                                                                                     [model.list dismissWithCompletion:NULL];
                                                                                                                 }
                                                                                                              isEnabled:YES];
    [startStopwatchAction setupAccessibilityWithAccessibilityLabel:@"Start Stopwatch"];
    [model addAction:startStopwatchAction];
    ActionsListDefaultButtonModel* startTimerAction = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Start Timer"
                                                                                                              image:[UIImage imageNamed:@"Timer"]
                                                                                                             action:^(ActionsListDefaultButtonModel* model) {
                                                                                                                 [model.list dismissWithCompletion:NULL];
                                                                                                             }
                                                                                                          isEnabled:YES];
    [startTimerAction setupAccessibilityWithAccessibilityLabel:@"Start Timer"];
    [model addAction:startTimerAction];
    
    [model presentWithCompletion:NULL];
    list = model;
}

- (void)presentFromNavigation:(UIButton*)button {
    ActionsListModel* model = [button createActionsListWithDelegate:NULL];
    [model setupDismissAccessibilityHint:@"Dismiss actions list"];
    ActionsListDefaultButtonModel* action1Model = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Test"
                                                                                                          image:[UIImage imageNamed:@"Dot"]
                                                                                                         action:^(ActionsListDefaultButtonModel* model) {
                                                                                                             [model.list dismissWithCompletion:^{
                                                                                                                 NSLog(@"Test");
                                                                                                             }];
                                                                                                         } isEnabled:YES];
    [action1Model setupAccessibilityWithAccessibilityLabel:@"Test accessibility label"];
    [model addAction:action1Model];
    
    ActionsListDefaultButtonModel* action2Model = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Test\nTest\nTest"
                                                                                                          image:[UIImage imageNamed:@"Dot"]
                                                                                                         action:^(ActionsListDefaultButtonModel* model) {
                                                                                                             [model.list dismissWithCompletion:^{
                                                                                                                 NSLog(@"Test");
                                                                                                             }];
                                                                                                         } isEnabled:YES];
    action2Model.appearance.font = [UIFont systemFontOfSize:10 weight:UIFontWeightBold];
    action2Model.appearance.highlightedBackgroundColor = UIColor.redColor;
    action2Model.appearance.highlightedTint = UIColor.yellowColor;
    action2Model.appearance.tint = UIColor.blueColor;
    action2Model.appearance.backgroundColor = UIColor.brownColor;
    [action2Model setupAccessibilityWithAccessibilityLabel:@"Triple test accessibility label"];
    [model addAction:action2Model];
    
    model.appearance.blurBackgroundColor = [UIColor.blackColor colorWithAlphaComponent:0.5];
    [model presentWithCompletion:NULL];
    list = model;
}

-(void)presentFromNavigationBar:(UIBarButtonItem*)barButton {
    ActionsListModel* model = [barButton createActionsListWithColor:[UIColor redColor]
                                                               font:NULL
                                                           delegate:NULL];
    [model setupDismissAccessibilityHint:@"Double tap to dismiss"];
    ActionsListDefaultButtonModel* action1Model = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Test\nTest"
                                                                                                          image:[UIImage imageNamed:@"Dot"]
                                                                                                         action:^(ActionsListDefaultButtonModel* model) {
                                                                                                             [model.list dismissWithCompletion:^{
                                                                                                                 NSLog(@"Test");
                                                                                                             }];
                                                                                                         } isEnabled:YES];
    [action1Model setupAccessibilityWithAccessibilityLabel:@"Double test accessibility label"];
    [model addAction:action1Model];
    
    ActionsListDefaultButtonModel* action2Model = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:@"Test\nTest\nTest\nTest"
                                                                                                          image:[UIImage imageNamed:@"Dot"]
                                                                                                         action:^(ActionsListDefaultButtonModel* model) {
                                                                                                             [model.list dismissWithCompletion:^{
                                                                                                                 NSLog(@"Test");
                                                                                                             }];
                                                                                                         } isEnabled:YES];
    [action2Model setupAccessibilityWithAccessibilityLabel:@"Quadruple test accessibility label"];
    [model addAction:action2Model];
                                                                                                             
    [model presentWithCompletion:NULL];
    list = model;
}

@end

File: Example/ActionsListDemoSwift/ViewController.swift
//
//  ActionsList
//  Copyright (c) 2017 LowKostKustomz. All rights reserved.
//

import UIKit
import ActionsList

class ViewController: UIViewController {
    
    private var list: ActionsListModel?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = nil
        createBackground()
        createNavigationItems()
        
        // Clocks button
        
        let button = UIButton(type: .system)
        button.setImage(UIImage(named: "Clock app")?.withRenderingMode(UIImageRenderingMode.alwaysOriginal), for: .normal)
        button.addTarget(self, action: #selector(clockButtonAction(sender:)), for: .touchUpInside)
        button.adjustsImageWhenHighlighted = false
        
        view.addSubview(button)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.heightAnchor.constraint(equalToConstant: 60).isActive = true
        button.widthAnchor.constraint(equalTo: button.heightAnchor).isActive = true
        
        button.centerXAnchor.constraint(equalTo: view.centerXAnchor, constant: -64).isActive = true
        
        button.centerYAnchor.constraint(equalTo: view.centerYAnchor, constant: -100).isActive = true
        button.setContentCompressionResistancePriority(UILayoutPriorityRequired, for: .horizontal)
    }
    
    private func createBackground() {
        let imageView = UIImageView(image: UIImage(named: "Wallpaper"))
        imageView.contentMode = .scaleAspectFill
        
        view.addSubview(imageView)
        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        imageView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        imageView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        imageView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
    }
    
    private func createNavigationItems() {
        let rightButton = UIButton(type: .custom)
        rightButton.setTitleColor(UIColor.black, for: .normal)
        rightButton.setTitle("Button", for: .normal)
        rightButton.sizeToFit()
        rightButton.addTarget(self, action: #selector(presentFromNavigation(button:)), for: .touchUpInside)
        
        let rightButton1 = UIButton(type: .custom)
        rightButton1.setImage(UIImage(named: "List icon")?.withRenderingMode(.alwaysTemplate), for: .normal)
        rightButton1.adjustsImageWhenHighlighted = false
        rightButton1.tintAdjustmentMode = .normal
        rightButton1.tintColor = UIColor.black
        rightButton1.sizeToFit()
        rightButton1.addTarget(self, action: #selector(presentFromNavigation(button:)), for: .touchUpInside)
        
        navigationItem.rightBarButtonItems = [UIBarButtonItem(customView: rightButton1),
                                              UIBarButtonItem(customView: rightButton)]
        
        navigationItem.leftBarButtonItems = [UIBarButtonItem(image: UIImage(named: "List icon"),
                                                             style: .plain,
                                                             target: self,
                                                             action: #selector(presentFromNavigation(barButton:))),
                                             UIBarButtonItem(title: "BarButton",
                                                             style: .plain,
                                                             target: self,
                                                             action: #selector(presentFromNavigation(barButton:)))]
    }
    
    @objc private func clockButtonAction(sender: UIButton) {
        let list = sender.createActionsList()
        
        list.add(action: ActionsListDefaultButtonModel(localizedTitle: "Create Alarm",
                                                       image: UIImage(named: "Alarm clock"),
                                                       action: { action in
                                                        action.list?.dismiss()
        }))
        list.add(action: ActionsListDefaultButtonModel(localizedTitle: "Start Stopwatch",
                                                       image: UIImage(named: "Stopwatch"),
                                                       action: { action in
                                                        action.list?.dismiss()
        }))
        list.add(action: ActionsListDefaultButtonModel(localizedTitle: "Start Timer",
                                                       image: UIImage(named: "Timer"),
                                                       action: { action in
                                                        action.list?.dismiss()
        }))
        
        list.present()
        self.list = list
    }
    
    @objc private func presentFromNavigation(button: UIButton) {
        let list = button.createActionsList()
        list.add(action: ActionsListDefaultButtonModel(localizedTitle: "Test",
                                                       image: UIImage(named: "Dot"),
                                                       action: { action in
                                                        action.list?.dismiss {
                                                            print("Test")
                                                        }
        }))
        
        let action2 =  ActionsListDefaultButtonModel(localizedTitle: "Test\nTest\nTest",
                                                     image: UIImage(named: "Dot"),
                                                     action: { action in
                                                        action.list?.reloadActions()
                                                        action.list?.dismiss {
                                                            print("Test")
                                                        }
        })
        action2.appearance.font = UIFont.systemFont(ofSize: 10, weight: UIFontWeightBold)
        action2.appearance.highlightedBackgroundColor = UIColor.red
        action2.appearance.highlightedTint = UIColor.yellow
        action2.appearance.tint = UIColor.blue
        action2.appearance.backgroundColor = UIColor.brown
        list.add(action: action2)
        
        list.appearance.blurBackgroundColor = UIColor.black.withAlphaComponent(0.5)
        list.present()
        self.list = list
    }
    
    @objc private func presentFromNavigation(barButton: UIBarButtonItem) {
        let list = barButton.createActionsList(withColor: UIColor.red,
                                               font: UIFont.systemFont(ofSize: 17, weight: UIFontWeightRegular))
        
        list?.add(action: ActionsListDefaultButtonModel(localizedTitle: "Test\nTest",
                                                        image: UIImage(named: "Dot"),
                                                        action: { action in
                                                            action.list?.dismiss {
                                                                print("Test")
                                                            }
        }))
        list?.add(action: ActionsListDefaultButtonModel(localizedTitle: "Test\nTest\nTest\nTest",
                                                        image: UIImage(named: "Dot"),
                                                        action: { action in
                                                            action.list?.dismiss {
                                                                print("Test")
                                                            }
        }))
        
        list?.present()
        self.list = list
    }
}

#if swift(>=4.0)
    private let UIFontWeightRegular = UIFont.Weight.regular
    private let UIFontWeightBold = UIFont.Weight.bold
    private let UILayoutPriorityRequired = UILayoutPriority.required
#endif
